src/util-lua-tls.c

```diff
@@ -316,48 +252,112 @@ static int GetCertChain(lua_State *luastate, const Flow *f, int direction)
+    if (connp->cert0_serial == NULL)
+        return LuaCallbackError(luastate, "error: no certificate serial");
 
-    Flow *f = LuaStateGetFlow(luastate);
-    if (f == NULL)
-        return LuaCallbackError(luastate, "internal error: no flow");
+    return LuaPushStringBuffer(
+            luastate, (uint8_t *)connp->cert0_serial, strlen(connp->cert0_serial));
+}
 
-    r = GetCertChain(luastate, f, direction);
+static int LuaTlsGetCertSerial(lua_State *luastate)
+{
+    struct LuaTls *s = (struct LuaTls *)lua_touserdata(luastate, 1);
+    if (s == NULL || s->f == NULL) {
+        LUA_ERROR("failed to get flow");
+    }
 
-    return r;
+    return GetCertSerial(luastate, s->f);
 }
 
-/** \brief register tls lua extensions in a luastate */
-int LuaRegisterTlsFunctions(lua_State *luastate)
+static int GetAgreedVersion(lua_State *luastate, Flow *f)
 {
-    /* registration of the callbacks */
-    lua_pushcfunction(luastate, TlsGetCertNotBefore);
-    lua_setglobal(luastate, "TlsGetCertNotBefore");
+    void *state = FlowGetAppState(f);
+    if (state == NULL)
+        return LuaCallbackError(luastate, "error: no app layer state");
+
+    SSLState *ssl_state = (SSLState *)state;
 
-    lua_pushcfunction(luastate, TlsGetCertNotAfter);
-    lua_setglobal(luastate, "TlsGetCertNotAfter");
+    int direction = LuaStateGetDirection(luastate);
+
+    int version;
+    if (direction) {
+        version = ssl_state->client_connp.version;
+    } else {
+        version = ssl_state->server_connp.version;
+    }
 
-    lua_pushcfunction(luastate, TlsGetVersion);
-    lua_setglobal(luastate, "TlsGetVersion");
+    char ssl_version[SSL_VERSION_MAX_STRLEN];
+    SSLVersionToString(version, ssl_version);
 
-    lua_pushcfunction(luastate, TlsGetCertInfo);
-    lua_setglobal(luastate, "TlsGetCertInfo");
+    lua_pushstring(luastate, (const char *)&ssl_version);
+    return 1;
+}
 
-    lua_pushcfunction(luastate, TlsGetSNI);
-    lua_setglobal(luastate, "TlsGetSNI");
+static int LuaTlsGetVersion(lua_State *luastate)
+{
+    struct LuaTls *s = (struct LuaTls *)lua_touserdata(luastate, 1);
+    if (s == NULL || s->f == NULL) {
+        LUA_ERROR("failed to get flow");
+    }
 
-    lua_pushcfunction(luastate, TlsGetCertSerial);
-    lua_setglobal(luastate, "TlsGetCertSerial");
+    return GetAgreedVersion(luastate, s->f);
+}
 
-    lua_pushcfunction(luastate, TlsGetCertChain);
-    lua_setglobal(luastate, "TlsGetCertChain");
+static const struct luaL_Reg tlslib_meta[] = {
+    // clang-format off
+    { "get_cert_not_before", LuaTlsGetCertNotBefore },
+    { "get_cert_not_after", LuaTlsGetCertNotAfter },
+    { "get_version", LuaTlsGetVersion },
+    { "get_serial", LuaTlsGetCertSerial },
+    { "get_cert_info", LuaTlsGetCertInfo },
+    { "get_sni", LuaTlsGetSNI },
+    { "get_cert_chain", LuaTlsGetCertChain },
+    { NULL, NULL, }
+    // clang-format on
+};
+
+static const struct luaL_Reg tlslib[] = {
+    // clang-format off
+    { "get", LuaTlsFlowGet },
```

Comment:
get_tx is what we use elsewhere for protocol libs

Reply:
You're suggesting `get_flow`?

Reply:
no `get_tx`

Reply:
Confused -- the TLS items are available from the flow's app layer state, not a tx.

Reply:
in tls, the state is exposed as a single tx